doc/design/error_reporting.md

```diff
@@ -165,7 +165,7 @@ the case that it does not hold
 ```cpp
 int x;
 // ...
-IOX_REQUIRE(x>=0, Code::OutOfBounds);
+IOX_REQUIRE(x>=0, "required condition violation message");
```

Comment:
@elBoberido I know it is merged but I think it is useful to have a code or more generic error object if it is to be passed to another framework (this is still the intention).

Some thing like

```IOX_REQUIRE(x>=0, Foo("Text", 73))```.

With sufficient support in the handler, anything can be passed and handled. In particular string literals, runtime constructed strings, objects of any kind, error wrappers around these. As of now, there exist no handling code for anything but codes (or string literals now). We decided this for simplicity.

It is important not to require to pass heavyweight objects, pointers to string literals or error codes are fine, but still allow more heavyweight objects if the user desires. Messages alone may not be descriptive enough at the handler site.

I would also liked to have the message (or any arguments) as optional, but variadic arguments do not play nice in some circumstances, but mainly if the case of no arguments must be supported (not necessarily the case here).

The semantics are also important. `IOX_REQUIRE` is not necessarily on a rare path in the "should not happen category (this would be `IOX_ASSUME`). Therefore performance is a consideration.

It may be a good idea to rethink whether we want more generic objects to be handled in the future. This should include string literals, codes and anything else. Fine for now and can be extended if we see the need.

Reply:
@MatthiasKillat for that there is still `IOX_REPORT_IF` and semantically it fits also better if it is intended to be forwarded. Currently I'm even thinking about simplifying the macros even more. Instead of having `IOX_PRECONDITION`, `IOX_ASSUME` and `IOX_REQUIRE` it could be `IOX_ASSERT` and `IOX_ENFORCE`. The former would be deactivated for release builds by default and nicely match the C assert. The latter would would emphasis the stronger requirements of always being active. With this the `IOX_REQUIRE` would again become available if we want to have it as a shortcut for `IOX_REPORT_IF`. We can discuss this further in the currently active PR.